File: leetcode/src/00981_time_based_key_value_store/solution.go
package leetcode

type TimeMap struct {
	keyTimestampMap map[string][]timestampVal
}

type timestampVal struct {
	timestamp int
	val       string
}

func Constructor() TimeMap {
	return TimeMap{
		keyTimestampMap: map[string][]timestampVal{},
	}
}

func (tm *TimeMap) Set(key string, value string, timestamp int) {
	tm.keyTimestampMap[key] = append(tm.keyTimestampMap[key], timestampVal{
		timestamp: timestamp,
		val:       value,
	})
}

func (tm *TimeMap) Get(key string, timestamp int) string {
	timestampVals, exist := tm.keyTimestampMap[key]
	if !exist {
		return ""
	}
	lo, hi := 0, len(timestampVals)
	for lo < hi {
		mid := lo + (hi-lo)/2
		if timestampVals[mid].timestamp > timestamp {
			hi = mid
		} else {
			lo = mid + 1
		}
	}
	if lo-1 < 0 {
		return ""
	}
	return timestampVals[lo-1].val
}

File: leetcode/src/00981_time_based_key_value_store/solution.rs
use std::collections::BTreeMap;
use std::collections::HashMap;

struct TimeMap {
    key_timestamp_map: HashMap<String, BTreeMap<i32, String>>,
}

impl TimeMap {
    fn new() -> Self {
        Self {
            key_timestamp_map: Default::default(),
        }
    }

    fn set(&mut self, key: String, value: String, timestamp: i32) {
        self.key_timestamp_map
            .entry(key)
            .or_default()
            .insert(timestamp, value);
    }

    fn get(&self, key: String, timestamp: i32) -> String {
        self.key_timestamp_map
            .get(&key)
            .and_then(|lookup| lookup.range(..=timestamp).next_back())
            .map(|(_, val)| val.to_owned())
            .unwrap_or_default()
    }
}
